ci/travis/bazel.py

```diff
@@ -0,0 +1,227 @@
+        self.extra_args = ("--show_progress=no", )
+
+    def _call(self, command, *args):
+        return subprocess.check_output(
+            self.argv + (command, ) + args[:1] + self.extra_args + args[1:],
+            stdin=subprocess.PIPE)
+
+    def info(self, *args):
+        result = OrderedDict()
+        for line in self._call("info", *args).splitlines():
+            (key, value) = line.split(b":", 1)
+            if value.startswith(b" "):
+                value = value[1:]
+            result[key.decode(self.encoding)] = value.decode(self.encoding)
+        return result
+
+    def aquery(self, *args):
+        lines = self._call("aquery", "--output=textproto", *args).splitlines()
+        return textproto_parse(lines, self.encoding, json.JSONEncoder())
+
+
+def parse_aquery_shell_calls(aquery_results):
+    for (key, val) in aquery_results:
+        if key == "actions":
+            [mnemonic] = [pair[1] for pair in val if pair[0] == "mnemonic"]
+            if mnemonic == "Genrule":
+                yield [pair[1] for pair in val if pair[0] == "arguments"]
+
+
+def parse_aquery_output_artifacts(aquery_results):
+    artifacts = {}
+    for (key, val) in aquery_results:
+        if key == "artifacts":
+            [artifact_id] = [pair[1] for pair in val if pair[0] == "id"]
+            [exec_path] = [pair[1] for pair in val if pair[0] == "exec_path"]
+            artifacts[artifact_id] = exec_path
+        elif key == "actions":
+            output_ids = [pair[1] for pair in val if pair[0] == "output_ids"]
+            for output_id in output_ids:
+                yield artifacts[output_id]
+
+
+def textproto2json(infile, outfile):
+    json_encoder = json.JSONEncoder(indent=2)
+    encoding = "utf-8"
+    for obj in textproto_parse(infile, encoding, json_encoder):
+        outfile.write((json_encoder.encode(obj) + "\n").encode(encoding))
+
+
+def preclean(bazel_aquery):
+    """Cleans up any genrule() outputs for the provided target(s).
+
+    This is useful for forcing genrule actions to re-run, because the _true_
+    outputs of those actions can include a larger set of files (e.g. files
+    copied to the workspace) which Bazel is unable to detect changes to (or
+    delete changes of).
+
+    Usually, you would run this script along with 'git clean -f', to make sure
+    Bazel re-copies outputs the next time a build occurs.
+    """
+    result = 0
+    bazel = Bazel()
+    aquery_results = bazel.aquery("--include_artifacts=true", bazel_aquery)
+    for path in parse_aquery_output_artifacts(aquery_results):
+        try:
+            if sys.platform == "win32":
+                os.chmod(path, stat.S_IWRITE)  # Needed to remove read-only bit
+            os.remove(path)
+        except IOError as ex:
+            if ex.errno != errno.ENOENT:
+                sys.stderr.write(str(ex) + "\n")
+                result = result or ex.errno
+    return result
+
+
+def shellcheck(bazel_aquery, *shellcheck_args):
```

Comment:
This file could really use some comments. What are you trying to do for shellcheck? What are some of the expected arguments?

Same with the other methods

Reply:
`shellcheck_args` is just arguments that get passed through to `shellcheck`. `bazel_aquery` is a Bazel aquery. I can try to mention these in comments.